src/dune_rules/artifact_substitution.ml

```diff
@@ -48,8 +48,22 @@ type conf =
   ; get_location : Section.t -> Package.Name.t -> Path.t
   ; get_config_path : configpath -> Path.t option
   ; hardcoded_ocaml_path : hardcoded_ocaml_path
+  ; post_copy_hook : (Path.t -> unit Fiber.t) option
   }
 
+let mac_codesign_hook ~codesign path =
+  Process.run Strict codesign [ "--sign"; "-"; Path.to_string path ]
+
+let post_copy_hook_of_context (context : Context.t) =
+  let config = context.ocaml_config in
+  match (Ocaml_config.system config, Ocaml_config.architecture config) with
+  | "macosx", "arm64" -> (
+    let codesign_name = "codesign" in
+    match Bin.which ~path:context.path codesign_name with
+    | None -> Utils.program_not_found ~loc:None codesign_name
+    | Some codesign -> Some (mac_codesign_hook ~codesign))
```

Comment:
Do we know that this identifies all the case of mac build that needs a new sign? macos has been shipping `codesign` for a long while, would it make sense to simply self-sign all binaries produced on `macos`?

Reply:
I'm more comfortable starting with a conservative solution and fix the problem only where it happens. As I understand it, gatekeeper enforces signatures on all binaries only on ARM so let's start with that. I agree that this is likely to evolve with new macos releases, we can adapt the test.